src/graphapi/engine.cpp

```diff
@@ -36,6 +43,179 @@ GraphPatternExecutor::~GraphPatternExecutor() = default;
 
 size_t GraphExecutorFind20::getWorkspaceSize() const { return mSolution.GetWorkspaceSize(); }
 
+namespace {
+
+const std::string_view base64Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                     "abcdefghijklmnopqrstuvwxyz"
+                                     "0123456789+/";
+
+std::string base64Encode(const char* data, size_t length)
+{
+    // Calculate the exact size of the resulting Base64 string
+    size_t outputSize = ((length + 2) / 3) * 4;
+    std::string encodedString;
+    encodedString.reserve(outputSize); // Preallocate memory for the result
+
+    size_t i = 0;
+    unsigned char charArray3[3];
+    unsigned char charArray4[4];
+
+    while(length-- != 0U)
+    {
+        charArray3[i++] = *(data++);
+        if(i == 3)
+        {
+            charArray4[0] = (charArray3[0] & 0xfc) >> 2;
+            charArray4[1] = ((charArray3[0] & 0x03) << 4) + ((charArray3[1] & 0xf0) >> 4);
+            charArray4[2] = ((charArray3[1] & 0x0f) << 2) + ((charArray3[2] & 0xc0) >> 6);
+            charArray4[3] = charArray3[2] & 0x3f;
+
+            std::transform(std::begin(charArray4),
+                           std::end(charArray4),
+                           std::back_inserter(encodedString),
+                           [](unsigned char c) { return base64Chars[c]; });
+            i = 0;
+        }
+    }
+
+    if(i != 0U)
+    {
+        for(size_t j = i; j < 3; j++)
+        {
+            charArray3[j] = '\0';
+        }
+
+        charArray4[0] = (charArray3[0] & 0xfc) >> 2;
+        charArray4[1] = ((charArray3[0] & 0x03) << 4) + ((charArray3[1] & 0xf0) >> 4);
+        charArray4[2] = ((charArray3[1] & 0x0f) << 2) + ((charArray3[2] & 0xc0) >> 6);
+        charArray4[3] = charArray3[2] & 0x3f;
+
+        for(size_t j = 0; j < i + 1; j++)
+        {
+            encodedString += base64Chars[charArray4[j]];
+        }
+
+        while(i++ < 3)
+        {
+            encodedString += '=';
+        }
+    }
+
+    return encodedString;
+}
+
+std::string base64Decode(const std::string& encodedString)
```

Comment:
Returning binary blob as std::string may have some problems on Windows, I guess @apwojcik has already fixed a few cases and replaced them with a std::vector<char>.

But probably it was related to a serialization of such strings via `ofstream`. Could you double check it?

Reply:
Agree. `std::vector` would be better here, regardless of whether `std::string` has issues with Windows. Even better `std::vector<uint8_t>`, though requires a `reinterpret_cast` for `miopenLoadSolution` and `miopenSaveSolution`